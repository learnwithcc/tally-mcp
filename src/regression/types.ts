export enum Severity {
    CRITICAL = 'critical',
    HIGH = 'high',
    MEDIUM = 'medium',
    LOW = 'low',
}

export interface RegressionTestCase {
    id: string; // e.g., 'BUG-123'
    description: string;
    severity: Severity;
    component: string;
    reproductionSteps: string[];
    expectedBehavior: string;
    // The actual test function
    test: () => Promise<void>;
}

export interface TestResult {
    testCaseId: string;
    passed: boolean;
    message?: string;
    duration: number;
} 