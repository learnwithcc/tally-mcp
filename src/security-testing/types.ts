/**
 * Security Testing Framework Types
 */

export type TestSuite = 'custom' | 'owasp-zap' | 'snyk';

export type SecurityTestStatus = 'passed' | 'failed' | 'warning' | 'skipped';

export type SecurityTestSeverity = 'low' | 'medium' | 'high' | 'critical';

export interface SecurityTestResult {
  id: string;
  name: string;
  suite: TestSuite;
  status: SecurityTestStatus;
  severity: SecurityTestSeverity;
  description: string;
  details?: string;
  evidence?: SecurityTestEvidence;
  remediation?: string;
  duration: number;
  timestamp: Date;
  metadata?: Record<string, any>;
}

export interface SecurityTestEvidence {
  type: 'request' | 'response' | 'log' | 'screenshot' | 'file';
  content: string;
  encoding?: 'base64' | 'utf8';
  mimeType?: string;
  size?: number;
}

export interface SecurityTestConfig {
  target: {
    baseUrl: string;
    apiKey?: string;
    headers?: Record<string, string>;
    timeout?: number;
  };
  suites: TestSuite[];
  owasp: OWASPZAPConfig;
  snyk: SnykConfig;
  custom: CustomTestConfig;
  reporting: ReportingConfig;
  schedule?: ScheduleConfig;
  cicd?: CICDConfig;
}

export interface OWASPZAPConfig {
  enabled: boolean;
  zapPath?: string;
  port?: number;
  host?: string;
  apiKey?: string;
  policies?: string[];
  scanTypes?: ('passive' | 'active' | 'baseline')[];
  excludeUrls?: string[];
  includeUrls?: string[];
  maxScanTime?: number;
  reportFormats?: ('html' | 'xml' | 'json' | 'md')[];
}

export interface SnykConfig {
  enabled: boolean;
  token?: string;
  orgId?: string;
  projectPath?: string;
  scanTypes?: ('code' | 'dependencies' | 'container' | 'iac')[];
  severity?: SecurityTestSeverity[];
  excludePatterns?: string[];
  failOnIssues?: boolean;
  monitorProject?: boolean;
}

export interface CustomTestConfig {
  enabled: boolean;
  testPaths?: string[];
  parallel?: boolean;
  maxConcurrency?: number;
  timeout?: number;
  retries?: number;
  skipPatterns?: string[];
}

export interface ReportingConfig {
  enabled: boolean;
  outputDir?: string;
  formats?: ('json' | 'html' | 'xml' | 'csv' | 'junit')[];
  includeEvidence?: boolean;
  aggregateResults?: boolean;
  webhooks?: WebhookConfig[];
  notifications?: NotificationConfig[];
}

export interface WebhookConfig {
  url: string;
  headers?: Record<string, string>;
  events?: ('test-start' | 'test-complete' | 'test-failure')[];
  template?: string;
}

export interface NotificationConfig {
  type: 'email' | 'slack' | 'teams' | 'webhook';
  target: string;
  events?: ('test-start' | 'test-complete' | 'test-failure')[];
  conditions?: {
    minSeverity?: SecurityTestSeverity;
    maxFailures?: number;
  };
}

export interface ScheduleConfig {
  enabled: boolean;
  cron?: string;
  timezone?: string;
  nextRun?: Date;
  autoStart?: boolean;
}

export interface CICDConfig {
  enabled: boolean;
  platform?: 'github' | 'gitlab' | 'jenkins' | 'azure-devops' | 'generic';
  breakBuildOnFailure?: boolean;
  publishResults?: boolean;
  artifactPath?: string;
  integrationKey?: string;
}

export interface SecurityTestSummary {
  startTime: number;
  endTime: number;
  duration: number;
  totalTests: number;
  passed: number;
  failed: number;
  warnings: number;
  skipped?: number;
  coverage?: {
    endpoints: number;
    vulnerabilities: Record<SecurityTestSeverity, number>;
  };
}

export interface SecurityTestMetrics {
  testExecutionTime: number;
  vulnerabilitiesFound: number;
  criticalIssues: number;
  highSeverityIssues: number;
  mediumSeverityIssues: number;
  lowSeverityIssues: number;
  falsePositives: number;
  coverage: number;
}

export interface SecurityTestReport {
  id: string;
  timestamp: Date;
  summary: SecurityTestSummary;
  results: SecurityTestResult[];
  metrics: SecurityTestMetrics;
  config: Partial<SecurityTestConfig>;
  environment: {
    platform: string;
    version: string;
    hostname: string;
    user: string;
  };
}

// Custom test interfaces for extensibility
export interface SecurityTest {
  id: string;
  name: string;
  description: string;
  severity: SecurityTestSeverity;
  category: string;
  tags: string[];
  execute(): Promise<SecurityTestResult>;
}

export interface SecurityTestCategory {
  name: string;
  description: string;
  tests: SecurityTest[];
  enabled: boolean;
}

export interface SecurityScanOptions {
  targetUrl: string;
  scanType?: 'passive' | 'active' | 'baseline';
  maxTime?: number;
  recursive?: boolean;
  excludeUrls?: string[];
  customHeaders?: Record<string, string>;
}

export interface VulnerabilityReport {
  id: string;
  title: string;
  description: string;
  severity: SecurityTestSeverity;
  cvss?: number;
  cve?: string;
  cwe?: string;
  solution?: string;
  references?: string[];
  affectedUrls?: string[];
  evidence?: SecurityTestEvidence[];
}

export interface SecurityTestRunner {
  initialize(): Promise<void>;
  runAll(): Promise<SecurityTestResult[]>;
  runTest(testId: string): Promise<SecurityTestResult>;
  cleanup(): Promise<void>;
  checkAvailability(): Promise<boolean>;
}

export interface SecurityTestIntegration extends SecurityTestRunner {
  name: string;
  version: string;
  configValid: boolean;
  runScan(options?: SecurityScanOptions): Promise<SecurityTestResult[]>;
} 