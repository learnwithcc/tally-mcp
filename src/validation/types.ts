/**
 * Type definitions for the MCP Tool Schema Validation Framework
 */

import { z } from 'zod';

/**
 * Severity levels for validation issues
 */
export enum ValidationSeverity {
  ERROR = 'error',
  WARNING = 'warning',
  INFO = 'info'
}

/**
 * Validation error structure
 */
export interface ValidationError {
  /** Unique identifier for the error type */
  code: string;
  /** Human-readable error message */
  message: string;
  /** Severity level of the validation issue */
  severity: ValidationSeverity;
  /** JSON path to the problematic field */
  path?: string;
  /** Expected value or format */
  expected?: any;
  /** Actual value found */
  actual?: any;
  /** Additional context about the error */
  context?: Record<string, any>;
  /** Reference to MCP specification section */
  specReference?: string;
}

/**
 * Result of a validation operation
 */
export interface ValidationResult {
  /** Whether the validation passed without errors */
  valid: boolean;
  /** List of validation errors found */
  errors: ValidationError[];
  /** List of warnings (non-blocking issues) */
  warnings: ValidationError[];
  /** Informational messages */
  info: ValidationError[];
  /** Total count of issues by severity */
  summary: {
    errorCount: number;
    warningCount: number;
    infoCount: number;
  };
}

/**
 * MCP Tool definition structure
 */
export interface MCPTool {
  /** Tool name */
  name: string;
  /** Tool description */
  description: string;
  /** JSON Schema for tool input parameters */
  inputSchema: Record<string, any>;
}

/**
 * Context information for tool validation
 */
export interface ToolValidationContext {
  /** The tool being validated */
  tool: MCPTool;
  /** Index of the tool in the tools array */
  toolIndex?: number | undefined;
  /** Additional validation options */
  options?: SchemaValidationOptions;
}

/**
 * Configuration options for schema validation
 */
export interface SchemaValidationOptions {
  /** Whether to perform strict validation */
  strict?: boolean;
  /** Whether to validate against specific MCP version */
  mcpVersion?: string;
  /** Whether to check for deprecated features */
  checkDeprecated?: boolean;
  /** Custom validation rules to apply */
  customRules?: ValidationRule[];
  /** Whether to validate JSON Schema draft version */
  validateJsonSchemaDraft?: boolean;
  /** Maximum allowed schema depth */
  maxSchemaDepth?: number;
}

/**
 * Custom validation rule interface
 */
export interface ValidationRule {
  /** Rule identifier */
  id: string;
  /** Rule description */
  description: string;
  /** Rule severity */
  severity: ValidationSeverity;
  /** Validation function */
  validate: (tool: MCPTool, context?: ToolValidationContext) => ValidationError[];
}

/**
 * Comprehensive validation report
 */
export interface ValidationReport {
  /** Timestamp of the validation */
  timestamp: string;
  /** Version of the validation framework */
  validatorVersion: string;
  /** MCP specification version used for validation */
  mcpVersion: string;
  /** Total number of tools validated */
  toolCount: number;
  /** Overall validation result */
  overall: ValidationResult;
  /** Individual tool validation results */
  toolResults: Array<{
    toolName: string;
    toolIndex: number;
    result: ValidationResult;
  }>;
  /** Performance metrics */
  metrics: {
    validationDurationMs: number;
    rulesExecuted: number;
    averageTimePerTool: number;
  };
  /** Configuration used for validation */
  config: SchemaValidationOptions;
}

/**
 * Zod schemas for runtime validation
 */
export const MCPToolSchema = z.object({
  name: z.string().min(1, 'Tool name cannot be empty'),
  description: z.string().min(1, 'Tool description cannot be empty'),
  inputSchema: z.record(z.any()).refine(
    (schema) => typeof schema === 'object' && schema !== null,
    'Input schema must be a valid object'
  )
});

export const ValidationErrorSchema = z.object({
  code: z.string(),
  message: z.string(),
  severity: z.nativeEnum(ValidationSeverity),
  path: z.string().optional(),
  expected: z.any().optional(),
  actual: z.any().optional(),
  context: z.record(z.any()).optional(),
  specReference: z.string().optional()
});

export const ValidationResultSchema = z.object({
  valid: z.boolean(),
  errors: z.array(ValidationErrorSchema),
  warnings: z.array(ValidationErrorSchema),
  info: z.array(ValidationErrorSchema),
  summary: z.object({
    errorCount: z.number(),
    warningCount: z.number(),
    infoCount: z.number()
  })
}); 