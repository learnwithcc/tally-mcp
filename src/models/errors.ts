/**
 * Standardized Error Classes for Tally API Client
 * 
 * These error classes provide structured error handling for different
 * HTTP status codes and API failure scenarios.
 */

/**
 * Base API error class that all other API errors extend
 */
export abstract class TallyApiError extends Error {
  public readonly isRetryable: boolean;
  public readonly statusCode?: number;
  public readonly statusText?: string;
  public readonly headers?: Record<string, string>;
  public readonly requestId?: string;
  public readonly timestamp: Date;
  public readonly cause?: Error;

  constructor(
    message: string,
    statusCode?: number,
    statusText?: string,
    headers?: Record<string, string>,
    isRetryable: boolean = false,
    cause?: Error
  ) {
    super(message);
    this.name = this.constructor.name;
    if (statusCode !== undefined) this.statusCode = statusCode;
    if (statusText !== undefined) this.statusText = statusText;
    if (headers !== undefined) this.headers = headers;
    this.isRetryable = isRetryable;
    this.timestamp = new Date();
    if (headers?.['x-request-id']) this.requestId = headers['x-request-id'];
    else if (headers?.['request-id']) this.requestId = headers['request-id'];
    if (cause !== undefined) this.cause = cause;
  }

  /**
   * Get a user-friendly error message
   */
  public getUserMessage(): string {
    return this.message;
  }

  /**
   * Get detailed error information for debugging
   */
  public getDebugInfo(): Record<string, any> {
    return {
      name: this.name,
      message: this.message,
      statusCode: this.statusCode,
      statusText: this.statusText,
      requestId: this.requestId,
      timestamp: this.timestamp.toISOString(),
      isRetryable: this.isRetryable,
      headers: this.headers,
    };
  }
}

/**
 * Authentication Error (401/403)
 * Thrown when authentication fails or token is invalid
 */
export class AuthenticationError extends TallyApiError {
  constructor(
    message: string = 'Authentication failed',
    statusCode?: number,
    statusText?: string,
    headers?: Record<string, string>,
    cause?: Error
  ) {
    super(message, statusCode, statusText, headers, false, cause);
  }

  public override getUserMessage(): string {
    if (this.statusCode === 401) {
      return 'Your session has expired. Please log in again.';
    }
    if (this.statusCode === 403) {
      return 'You do not have permission to access this resource.';
    }
    return 'Authentication failed. Please check your credentials.';
  }
}

/**
 * Bad Request Error (400)
 * Thrown when the request is malformed or invalid
 */
export class BadRequestError extends TallyApiError {
  public readonly validationErrors?: Array<{
    field?: string;
    message: string;
    code?: string;
  }>;

  constructor(
    message: string = 'Bad request',
    statusCode: number = 400,
    statusText?: string,
    headers?: Record<string, string>,
    validationErrors?: Array<{ field?: string; message: string; code?: string }>,
    cause?: Error
  ) {
    super(message, statusCode, statusText, headers, false, cause);
    if (validationErrors !== undefined) this.validationErrors = validationErrors;
  }

  public override getUserMessage(): string {
    if (this.validationErrors && this.validationErrors.length > 0) {
      const fieldErrors = this.validationErrors
        .filter(error => error.field)
        .map(error => `${error.field}: ${error.message}`)
        .join(', ');
      
      if (fieldErrors) {
        return `Validation failed: ${fieldErrors}`;
      }
    }
    return 'The request contains invalid data. Please check your input and try again.';
  }
}

/**
 * Not Found Error (404)
 * Thrown when the requested resource is not found
 */
export class NotFoundError extends TallyApiError {
  constructor(
    message: string = 'Resource not found',
    statusCode: number = 404,
    statusText?: string,
    headers?: Record<string, string>,
    cause?: Error
  ) {
    super(message, statusCode, statusText, headers, false, cause);
  }

  public override getUserMessage(): string {
    return 'The requested resource was not found.';
  }
}

/**
 * Rate Limit Error (429)
 * Thrown when rate limits are exceeded
 */
export class RateLimitError extends TallyApiError {
  public readonly retryAfter?: number;
  public readonly limit?: number;
  public readonly remaining?: number;
  public readonly resetTime?: Date;

  constructor(
    message: string = 'Rate limit exceeded',
    statusCode: number = 429,
    statusText?: string,
    headers?: Record<string, string>,
    cause?: Error
  ) {
    super(message, statusCode, statusText, headers, true, cause);
    
    // Parse rate limit headers
    if (headers) {
      if (headers['retry-after']) {
        const retryAfterParsed = parseInt(headers['retry-after'], 10);
        if (!isNaN(retryAfterParsed)) this.retryAfter = retryAfterParsed;
      }
      if (headers['x-ratelimit-limit']) {
        const limitParsed = parseInt(headers['x-ratelimit-limit'], 10);
        if (!isNaN(limitParsed)) this.limit = limitParsed;
      }
      if (headers['x-ratelimit-remaining']) {
        const remainingParsed = parseInt(headers['x-ratelimit-remaining'], 10);
        if (!isNaN(remainingParsed)) this.remaining = remainingParsed;
      }
      if (headers['x-ratelimit-reset']) {
        const resetTimeParsed = parseInt(headers['x-ratelimit-reset'], 10);
        if (!isNaN(resetTimeParsed)) this.resetTime = new Date(resetTimeParsed * 1000);
      }
    }
  }

  public override getUserMessage(): string {
    if (this.retryAfter) {
      return `Rate limit exceeded. Please wait ${this.retryAfter} seconds before trying again.`;
    }
    if (this.resetTime) {
      return `Rate limit exceeded. Limit resets at ${this.resetTime.toLocaleTimeString()}.`;
    }
    return 'Rate limit exceeded. Please try again later.';
  }
}

/**
 * Server Error (500+)
 * Thrown when the server encounters an internal error
 */
export class ServerError extends TallyApiError {
  constructor(
    message: string = 'Internal server error',
    statusCode: number = 500,
    statusText?: string,
    headers?: Record<string, string>,
    cause?: Error
  ) {
    // Server errors are typically retryable
    super(message, statusCode, statusText, headers, true, cause);
  }

  public override getUserMessage(): string {
    if (this.statusCode === 502) {
      return 'Service temporarily unavailable. Please try again later.';
    }
    if (this.statusCode === 503) {
      return 'Service is currently under maintenance. Please try again later.';
    }
    if (this.statusCode === 504) {
      return 'Request timed out. Please try again.';
    }
    return 'A server error occurred. Please try again later.';
  }
}

/**
 * Network Error
 * Thrown when network connectivity issues occur
 */
export class NetworkError extends TallyApiError {
  constructor(
    message: string = 'Network error',
    cause?: Error
  ) {
    super(message, undefined, undefined, undefined, true, cause);
  }

  public override getUserMessage(): string {
    return 'Network connection failed. Please check your internet connection and try again.';
  }
}

/**
 * Timeout Error
 * Thrown when requests exceed the configured timeout
 */
export class TimeoutError extends TallyApiError {
  constructor(
    message: string = 'Request timeout',
    cause?: Error
  ) {
    super(message, undefined, undefined, undefined, true, cause);
  }

  public override getUserMessage(): string {
    return 'Request timed out. Please try again.';
  }
}

/**
 * Factory function to create appropriate error based on HTTP status code
 */
export function createErrorFromResponse(
  statusCode: number,
  statusText: string,
  headers: Record<string, string>,
  responseData?: any,
  originalError?: Error
): TallyApiError {
  const message = responseData?.message || responseData?.error || statusText || 'Unknown error';
  
  switch (statusCode) {
    case 400:
      return new BadRequestError(
        message,
        statusCode,
        statusText,
        headers,
        responseData?.errors || responseData?.validation_errors,
        originalError
      );
    
    case 401:
    case 403:
      return new AuthenticationError(message, statusCode, statusText, headers, originalError);
    
    case 404:
      return new NotFoundError(message, statusCode, statusText, headers, originalError);
    
    case 429:
      return new RateLimitError(message, statusCode, statusText, headers, originalError);
    
    default:
      if (statusCode >= 500) {
        return new ServerError(message, statusCode, statusText, headers, originalError);
      }
      
      // For other 4xx errors, use BadRequestError as fallback
      if (statusCode >= 400 && statusCode < 500) {
        return new BadRequestError(message, statusCode, statusText, headers, undefined, originalError);
      }
      
      // Fallback for unexpected status codes
      return new ServerError(message, statusCode, statusText, headers, originalError);
  }
} 