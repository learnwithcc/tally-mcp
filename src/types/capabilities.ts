/**
 * MCP Server Capability Types
 */

/**
 * Base capability interface with common properties
 */
interface BaseCapability {
  listChanged?: boolean;
  subscribe?: boolean;
}

/**
 * Tools capability configuration
 */
interface ToolsCapability extends BaseCapability {
  // Tool-specific capabilities
  call?: boolean;
  list?: boolean;
}

/**
 * Resources capability configuration
 */
interface ResourcesCapability extends BaseCapability {
  // Resource-specific capabilities
  get?: boolean;
  put?: boolean;
  delete?: boolean;
}

/**
 * Prompts capability configuration
 */
interface PromptsCapability extends BaseCapability {
  // Prompt-specific capabilities
  get?: boolean;
  list?: boolean;
}

/**
 * Logging capability configuration
 */
interface LoggingCapability {
  level?: 'debug' | 'info' | 'warn' | 'error';
  subscribe?: boolean;
}

/**
 * Server capabilities interface
 */
export interface ServerCapabilities {
  protocolVersion: string;
  tools?: ToolsCapability;
  resources?: ResourcesCapability;
  prompts?: PromptsCapability;
  logging?: LoggingCapability;
}

/**
 * Client capabilities interface
 */
export interface ClientCapabilities {
  protocolVersion: string;
  tools?: Partial<ToolsCapability>;
  resources?: Partial<ResourcesCapability>;
  prompts?: Partial<PromptsCapability>;
  logging?: Partial<LoggingCapability>;
}

/**
 * Negotiated capabilities between server and client
 */
export interface NegotiatedCapabilities {
  protocolVersion: string;
  tools?: ToolsCapability;
  resources?: ResourcesCapability;
  prompts?: PromptsCapability;
  logging?: LoggingCapability;
}

/**
 * Default server capabilities
 */
export const DEFAULT_SERVER_CAPABILITIES: ServerCapabilities = {
  protocolVersion: '2024-11-05',
  tools: {
    call: true,
    list: true,
    listChanged: true,
    subscribe: true
  },
  resources: {
    get: true,
    put: true,
    delete: true,
    listChanged: true,
    subscribe: true
  },
  prompts: {
    get: true,
    list: true,
    listChanged: true,
    subscribe: true
  },
  logging: {
    level: 'info',
    subscribe: true
  }
}; 