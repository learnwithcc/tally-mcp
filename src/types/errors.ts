/**
 * Structured error interface for consistent error handling
 */
export interface StructuredErrorData {
  code: string;
  message: string;
  data?: Record<string, unknown>;
}

/**
 * Custom error class for structured error handling
 */
export class StructuredError extends Error {
  public readonly code: string;
  public readonly data: Record<string, unknown> | undefined;

  constructor({ code, message, data }: StructuredErrorData) {
    super(message);
    this.name = 'StructuredError';
    this.code = code;
    this.data = data;

    // Ensure proper prototype chain for instanceof checks
    Object.setPrototypeOf(this, StructuredError.prototype);
  }
} 