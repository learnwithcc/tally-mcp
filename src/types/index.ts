export * from './monitoring';
export * from './capabilities';
export * from './errors';
export * from './json-rpc'; 