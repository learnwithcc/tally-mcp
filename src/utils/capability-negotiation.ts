import { ServerCapabilities, ClientCapabilities, NegotiatedCapabilities, DEFAULT_SERVER_CAPABILITIES } from '../types/capabilities';
import { StructuredError } from '../types/errors';

/**
 * Check if protocol versions are compatible
 * @param serverVersion Server protocol version
 * @param clientVersion Client protocol version
 * @returns True if versions are compatible
 */
function areVersionsCompatible(serverVersion: string, clientVersion: string): boolean {
  // For now, versions must match exactly. In the future, we can implement semver comparison
  return serverVersion === clientVersion;
}

/**
 * Negotiate capabilities between server and client
 * @param clientCapabilities Optional client capabilities to negotiate with
 * @returns Negotiated capabilities
 * @throws {StructuredError} If protocol versions are incompatible
 */
export function negotiateCapabilities(clientCapabilities?: ClientCapabilities): NegotiatedCapabilities {
  // Start with protocol version
  const negotiatedCapabilities: NegotiatedCapabilities = {
    protocolVersion: DEFAULT_SERVER_CAPABILITIES.protocolVersion
  };

  // If no client capabilities provided, copy server defaults
  if (!clientCapabilities) {
    if (DEFAULT_SERVER_CAPABILITIES.tools) {
      negotiatedCapabilities.tools = { ...DEFAULT_SERVER_CAPABILITIES.tools };
    }
    if (DEFAULT_SERVER_CAPABILITIES.resources) {
      negotiatedCapabilities.resources = { ...DEFAULT_SERVER_CAPABILITIES.resources };
    }
    if (DEFAULT_SERVER_CAPABILITIES.prompts) {
      negotiatedCapabilities.prompts = { ...DEFAULT_SERVER_CAPABILITIES.prompts };
    }
    if (DEFAULT_SERVER_CAPABILITIES.logging) {
      negotiatedCapabilities.logging = { ...DEFAULT_SERVER_CAPABILITIES.logging };
    }
    return negotiatedCapabilities;
  }

  // Check protocol version compatibility
  if (!areVersionsCompatible(DEFAULT_SERVER_CAPABILITIES.protocolVersion, clientCapabilities.protocolVersion)) {
    throw new StructuredError({
      code: 'INCOMPATIBLE_PROTOCOL_VERSION',
      message: `Protocol version mismatch. Server: ${DEFAULT_SERVER_CAPABILITIES.protocolVersion}, Client: ${clientCapabilities.protocolVersion}`,
      data: {
        serverVersion: DEFAULT_SERVER_CAPABILITIES.protocolVersion,
        clientVersion: clientCapabilities.protocolVersion
      }
    });
  }

  // Negotiate tools capabilities
  if (clientCapabilities.tools && DEFAULT_SERVER_CAPABILITIES.tools) {
    negotiatedCapabilities.tools = {
      call: clientCapabilities.tools.call ?? DEFAULT_SERVER_CAPABILITIES.tools.call ?? false,
      list: clientCapabilities.tools.list ?? DEFAULT_SERVER_CAPABILITIES.tools.list ?? false,
      listChanged: clientCapabilities.tools.listChanged ?? DEFAULT_SERVER_CAPABILITIES.tools.listChanged ?? false,
      subscribe: clientCapabilities.tools.subscribe ?? DEFAULT_SERVER_CAPABILITIES.tools.subscribe ?? false
    };
  }

  // Negotiate resources capabilities
  if (clientCapabilities.resources && DEFAULT_SERVER_CAPABILITIES.resources) {
    negotiatedCapabilities.resources = {
      get: clientCapabilities.resources.get ?? DEFAULT_SERVER_CAPABILITIES.resources.get ?? false,
      put: clientCapabilities.resources.put ?? DEFAULT_SERVER_CAPABILITIES.resources.put ?? false,
      delete: clientCapabilities.resources.delete ?? DEFAULT_SERVER_CAPABILITIES.resources.delete ?? false,
      listChanged: clientCapabilities.resources.listChanged ?? DEFAULT_SERVER_CAPABILITIES.resources.listChanged ?? false,
      subscribe: clientCapabilities.resources.subscribe ?? DEFAULT_SERVER_CAPABILITIES.resources.subscribe ?? false
    };
  }

  // Negotiate prompts capabilities
  if (clientCapabilities.prompts && DEFAULT_SERVER_CAPABILITIES.prompts) {
    negotiatedCapabilities.prompts = {
      get: clientCapabilities.prompts.get ?? DEFAULT_SERVER_CAPABILITIES.prompts.get ?? false,
      list: clientCapabilities.prompts.list ?? DEFAULT_SERVER_CAPABILITIES.prompts.list ?? false,
      listChanged: clientCapabilities.prompts.listChanged ?? DEFAULT_SERVER_CAPABILITIES.prompts.listChanged ?? false,
      subscribe: clientCapabilities.prompts.subscribe ?? DEFAULT_SERVER_CAPABILITIES.prompts.subscribe ?? false
    };
  }

  // Negotiate logging capabilities
  if (clientCapabilities.logging && DEFAULT_SERVER_CAPABILITIES.logging) {
    negotiatedCapabilities.logging = {
      level: clientCapabilities.logging.level ?? DEFAULT_SERVER_CAPABILITIES.logging.level ?? 'info',
      subscribe: clientCapabilities.logging.subscribe ?? DEFAULT_SERVER_CAPABILITIES.logging.subscribe ?? false
    };
  }

  return negotiatedCapabilities;
}

/**
 * Validate client capabilities against server capabilities
 * @param capabilities Client capabilities to validate
 * @returns True if capabilities are valid, false otherwise
 */
export function validateClientCapabilities(capabilities: unknown): capabilities is ClientCapabilities {
  if (!capabilities || typeof capabilities !== 'object') {
    return false;
  }

  const capObj = capabilities as Record<string, unknown>;
  
  // Check for unknown capability categories
  const validCategories = ['tools', 'resources', 'prompts', 'logging'];
  const hasInvalidCategory = Object.keys(capObj).some(key => !validCategories.includes(key));
  
  if (hasInvalidCategory) {
    return false;
  }

  // Validate tools capabilities
  if (capObj.tools && typeof capObj.tools !== 'object') {
    return false;
  }

  // Validate resources capabilities
  if (capObj.resources && typeof capObj.resources !== 'object') {
    return false;
  }

  // Validate prompts capabilities
  if (capObj.prompts && typeof capObj.prompts !== 'object') {
    return false;
  }

  // Validate logging capabilities
  if (capObj.logging) {
    if (typeof capObj.logging !== 'object') {
      return false;
    }
    
    const logging = capObj.logging as Record<string, unknown>;
    if (logging.level && !['debug', 'info', 'warn', 'error'].includes(logging.level as string)) {
      return false;
    }
  }

  return true;
} 