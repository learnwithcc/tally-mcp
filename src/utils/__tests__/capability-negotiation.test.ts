import { negotiateCapabilities, validateClientCapabilities } from '../capability-negotiation';
import { ClientCapabilities, DEFAULT_SERVER_CAPABILITIES } from '../../types/capabilities';
import { StructuredError } from '../../types/errors';
import { capabilityNegotiationTestCases, getMockClientCapabilities } from './fixtures';

describe('Capability Negotiation', () => {

  describe('negotiateCapabilities', () => {
    
    it('should return server defaults when no client capabilities are provided', () => {
      const negotiated = negotiateCapabilities();
      expect(negotiated).toEqual(DEFAULT_SERVER_CAPABILITIES);
    });

    it('should throw an error for incompatible protocol versions', () => {
      const clientCaps: ClientCapabilities = {
        protocolVersion: '0.1.0', // Different from server default
      };
      expect(() => negotiateCapabilities(clientCaps)).toThrow(StructuredError);
      try {
        negotiateCapabilities(clientCaps);
      } catch (e) {
        const error = e as StructuredError;
        expect(error.code).toBe('INCOMPATIBLE_PROTOCOL_VERSION');
        expect(error.data).toEqual({
          serverVersion: DEFAULT_SERVER_CAPABILITIES.protocolVersion,
          clientVersion: clientCaps.protocolVersion,
        });
        expect(error).toMatchSnapshot();
      }
    });

    test.each(capabilityNegotiationTestCases)(
      '$description',
      ({ clientCaps, expected }) => {
        const negotiated = negotiateCapabilities(clientCaps);
        expect(negotiated).toEqual(expected);
      },
    );

    it('should use server default when a client capability is not specified', () => {
      const clientCaps: ClientCapabilities = getMockClientCapabilities({
        tools: {}, // Providing the category but not the specific capability
      });

      const negotiated = negotiateCapabilities(clientCaps);
      // It should fallback to the server's default for 'call'
      expect(negotiated.tools?.call).toBe(DEFAULT_SERVER_CAPABILITIES.tools?.call);
      expect(negotiated).toMatchSnapshot();
    });

  });

  describe('validateClientCapabilities', () => {
    
    const validCases: [string, object][] = [
      ['full capabilities', { tools: { call: true }, logging: { level: 'warn', subscribe: false } }],
      ['partial capabilities', { tools: { call: true } }],
      ['empty capabilities', {}],
    ];

    test.each(validCases)('should return true for valid capabilities: %s', (_, caps) => {
      expect(validateClientCapabilities(caps)).toBe(true);
    });

    const invalidCases: [string, unknown][] = [
      ['null', null],
      ['string', 'string'],
      ['number', 123],
      ['unknown category', { unknownCategory: { read: true } }],
      ['category not an object', { tools: 'not-an-object' }],
      ['invalid logging level', { logging: { level: 'invalid-level' } }],
    ];

    test.each(invalidCases)('should return false for invalid capabilities: %s', (_, caps) => {
      expect(validateClientCapabilities(caps)).toBe(false);
    });

  });
}); 