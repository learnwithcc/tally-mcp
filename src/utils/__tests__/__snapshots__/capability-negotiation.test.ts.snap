// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Capability Negotiation negotiateCapabilities should throw an error for incompatible protocol versions 1`] = `[StructuredError: Protocol version mismatch. Server: 2024-11-05, Client: 0.1.0]`;

exports[`Capability Negotiation negotiateCapabilities should use server default when a client capability is not specified 1`] = `
{
  "logging": {
    "level": "info",
    "subscribe": true,
  },
  "prompts": {
    "get": true,
    "list": true,
    "listChanged": true,
    "subscribe": true,
  },
  "protocolVersion": "2024-11-05",
  "resources": {
    "delete": true,
    "get": true,
    "listChanged": true,
    "put": true,
    "subscribe": true,
  },
  "tools": {
    "call": true,
    "list": true,
    "listChanged": true,
    "subscribe": true,
  },
}
`;
